dali/pipeline/executor/executor.h

```diff
@@ -140,6 +140,51 @@ class DLL_PUBLIC Executor : public ExecutorBase, public WorkspacePolicy, public
 
   void SetupOutputQueuesForGraph();
 
+  template <typename Workspace>
+  void RunHelper(OperatorBase &op, Workspace &ws) {
+    constexpr size_t stage_id = static_cast<size_t>(workspace_to_op<Workspace>::value);
+    output_desc_[stage_id].clear();
+    if (op.Setup(output_desc_[stage_id], ws)) {
+      DALI_ENFORCE(
+          static_cast<size_t>(ws.NumOutput()) == output_desc_[stage_id].size(),
+          "Operator::Setup returned shape and type information for mismatched number of outputs");
+      for (int i = 0; i < ws.NumOutput(); i++) {
+        auto &desc = output_desc_[stage_id][i];
+        if (ws.template OutputIsType<CPUBackend>(i)) {
+          ws.template OutputRef<CPUBackend>(i).Resize(desc.shape);
+          ws.template OutputRef<CPUBackend>(i).set_type(desc.type);
+        } else {
+          ws.template OutputRef<GPUBackend>(i).Resize(desc.shape);
+          ws.template OutputRef<GPUBackend>(i).set_type(desc.type);
+        }
+      }
+    }
+    op.Run(&ws);
+  }
+
+  void RunHelper(OperatorBase &op, SupportWorkspace &ws) {
+    size_t stage_id = static_cast<size_t>(OpType::SUPPORT);
+    output_desc_[stage_id].clear();
```

Comment:
I don't like `output_desc_` being stored per stage. It has potential to break when we have parallel execution of setup. There are two other elegant solutions and one dirty (keeping them as `static thread_local`):
* Create a `RunHelper` _object_ that will have template function `Run` or `operator()` and store output_descs. If RunHelper is kept locally by a thread that runs Setup, thread safety is maintained. Disadvantage: the vectors are lost in each Run.
* Store output_desc per node - full parallelism, persistent object in steady state. Disadvantage: memory consumption (but it's just shapes, so not that bad).

Reply:
I moved it to OpGraph, so it is stored per-op now.